### PR TITLE
Refactor boavizta model

### PR DESCRIPTION
**Pull request by manushak**
### Types of changes

- [x] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### A description of the changes proposed in the Pull Request

- Refactored boavizta model and tests
-

### CODE REVIEW

File: src/lib/azure-importer/README.md [hunk 1]
```diff
@@ -161,5 +152,14 @@ outputs:
...
     cloud-instance-type: Standard_B1s
-...
 ```
+
+You can run this example `impl` by saving it as `./examples/impls/test/azure-importer.yml` and executing the following command from the project root:
+
+```sh
+npm i -g @grnsft/if
+npm i -g @grnsft/if-unofficial-models
+impact-engine --impl ./examples/impls/test/azure-importer.yml --ompl ./examples/ompls/azure-importer.yml
```
Comment (narekhovhannisyan) [inline src/lib/azure-importer/README.md:162]: doc is for azure importer

File: src/lib/boavizta/base-output-model.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      result.push(usageResult);
+    }
+
+    return result;
+  }
+
+  /**
+   * Fetches data from Boavizta API according to the specific endpoint of the model
+   */
+  protected abstract fetchData(usageData: object | undefined): Promise<object>;
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/base-output-model.ts:61]: generic can be used here instead

File: src/lib/boavizta/base-output-model.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * Abstract subs to make compatibility with base interface. allows configure to be defined in base class
+   */
+  protected abstract captureStaticParams(staticParams: object): Promise<object>;
+
+  /**
+   * Extracts information from Boavizta API response to return the output in the format required by IMPL.
+   */
+  protected formatResponse(data: KeyValuePair) {
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/base-output-model.ts:99]:
```ts
  protected formatResponse(data: KeyValuePair) {
    const impactsInData = 'impacts' in data;
    const embodiedCarbon = impactsInData
      ? data.impacts.gwp.embedded.value * 1000
      : 0;
    const energy = impactsInData ? data.impacts.pe.use.value / 3.6 : 0;

    return {'embodied-carbon': embodiedCarbon, energy};
  }
```

Comment (narekhovhannisyan) [reply src/lib/boavizta/base-output-model.ts:99]: docs can be introduced in JSDoc on the top of the function

File: src/lib/boavizta/base-output-model.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        })
+      );
+    }
+
+    const usageInput = this.transformToBoaviztaUsage(
+      input['duration'],
+      input[this.metricType]
+    );
+
+    return (await this.fetchData(usageInput)) as IBoaviztaUsageSCI;
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/base-output-model.ts:132]: if generic will be used for fetchData, you don't have to await to it

File: src/lib/boavizta/boavizta-api.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * Fetches cloud instance data from Boavizta API.
+   */
+  public async fetchCloudInstanceData(
+    dataCast: KeyValuePair,
+    verbose: boolean
+  ): Promise<object> {
+    // Replace '-' with '_' in keys
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/boavizta-api.ts:31]: unnecessary doc

File: src/lib/boavizta/boavizta-api.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    );
+
+    return instances.data;
+  }
+
+  /**
+   * Gets the list of supported cloud providers.
+   */
+  public async getSupportedProvidersList(): Promise<string[]> {
+    const providers = await axios.get(
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/boavizta-api.ts:56]: axios supports generics, if you know the type on response, you can inject it

File: src/lib/boavizta/index.ts [hunk 7]
```diff
@@ -1,161 +1,52 @@
...
-    };
-    // convert expected lifespan from seconds to years
-    usageInput['years_life_time'] =
-      this.expectedLifespan / (365.0 * 24.0 * 60.0 * 60.0);
-    usageInput = this.addLocationToUsage(usageInput);
+export class BoaviztaCpuOutputModel
+  extends BoaviztaBaseOutputModel
+  implements ModelPluginInterface
+{
+  public allocation = 'LINEAR';
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/index.ts:18]: let's check if this should be public

File: src/lib/boavizta/index.ts [hunk 8]
```diff
@@ -164,219 +55,178 @@ export class BoaviztaCpuOutputModel
...
+    await this.validateLocation(safeStaticParams);
 
-    if (!('instance-type' in staticParamsCast)) {
-      throw new InputValidationError(
-        this.errorBuilder({
-          message: "Missing 'instance-type' parameter from configuration",
-        })
-      );
+    if ('expected-lifespan' in safeStaticParams) {
+      this.expectedLifespan = safeStaticParams['expected-lifespan'] as number;
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/index.ts:117]: please check if explicitly assigning type as a number needed

File: src/lib/boavizta/index.ts [hunk 9]
```diff
@@ -164,219 +55,178 @@ export class BoaviztaCpuOutputModel
...
-  async fetchData(usageData: object | undefined): Promise<object> {
-    if (this.sharedParams === undefined) {
+  /**
+   * Validates the provider parameter for the Cloud model.
+   */
+  private async validateProvider(staticParamsCast: ModelParams) {
+    const supportedProviders =
+      await this.boaviztaAPI.getSupportedProvidersList();
+
+    if (!supportedProviders.includes(staticParamsCast.provider as string)) {
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/index.ts:171]: if possible please reduce usage of explicit assignment

File: src/lib/boavizta/index.ts [hunk 10]
```diff
@@ -164,219 +55,178 @@ export class BoaviztaCpuOutputModel
...
-      }
     }
-  }
 
-  async supportedInstancesList(provider: string) {
-    const instances = await axios.get(
-      `https://api.boavizta.org/v1/cloud/instance/all_instances?provider=${provider}`
-    );
+    const dataCast: object = Object.assign({}, this.sharedParams, {
+      usage: usageData,
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/index.ts:138]: usage and usage data can share the same name

File: src/lib/boavizta/index.ts [hunk 11]
```diff
@@ -164,219 +61,178 @@ export class BoaviztaCpuOutputModel
...
+  private async validateLocation(
+    staticParamsCast: ModelParams
+  ): Promise<string | void> {
+    if ('location' in staticParamsCast) {
+      const location = staticParamsCast.location as string;
+      const countries = await this.boaviztaAPI.getSupportedLocations();
 
-    await this.validateLocation(staticParams); // if no valid location found, throw error
+      if (!countries.includes(location)) {
+        throw new InputValidationError(
```
Comment (narekhovhannisyan) [inline src/lib/boavizta/index.ts:226]: can be like this to make error message lighter
```
const whitelistedCountries = countries.join(', ');

        throw new InputValidationError(
          this.errorBuilder({
            message: `Invalid location parameter location. Valid values are ${whitelistedCountries}`,
          })
        );
```

Comment (manushak) [reply src/lib/boavizta/index.ts:219]: done

File: src/lib/boavizta/index.ts (hunk 10 above)
Comment (manushak) [reply src/lib/boavizta/index.ts]: done

File: src/lib/boavizta/index.ts (hunk 9 above)
Comment (manushak) [reply src/lib/boavizta/index.ts]: done

File: src/lib/boavizta/index.ts (hunk 8 above)
Comment (manushak) [reply src/lib/boavizta/index.ts]: done

File: src/lib/boavizta/index.ts (hunk 7 above)
Comment (manushak) [reply src/lib/boavizta/index.ts]: done

File: src/lib/boavizta/boavizta-api.ts (hunk 6 above)
Comment (manushak) [reply src/lib/boavizta/boavizta-api.ts]: done

File: src/lib/boavizta/boavizta-api.ts (hunk 5 above)
Comment (manushak) [reply src/lib/boavizta/boavizta-api.ts]: removed

File: src/lib/boavizta/base-output-model.ts (hunk 4 above)
Comment (manushak) [reply src/lib/boavizta/base-output-model.ts]: changed

**Review by narekhovhannisyan — APPROVED** [review]

**Pull request merged**